File: What the find/C#/Extensions.cs
using System;
using System.Collections.Generic;

namespace WhatTheFind
{
    public static class Extensions
    {
        /// <summary>
        /// Find any node in an object graph that satisfy a given predicate and return it.
        /// </summary>
        /// <typeparam name="T">Type of object.</typeparam>
        /// <param name="root">The root node.</param>
        /// <param name="predicate">The given condition to satisfy.</param>
        /// <param name="getChildren">Child selector.</param>
        /// <returns>Node satisfying the condtion, else null.</returns>
        public static T FindWhere<T>(this T root, Func<T, bool> predicate, Func<T, IEnumerable<T>> getChildren)
            where T : class
        {
            // YOUR SOLUTION GOES HERE
            throw new NotImplementedException();
        }
    }
}

File: What the find/Java/src/com/itminds/Extensions.java
package com.itminds;

import java.util.function.Predicate;
import java.util.function.Function;

/**
 *  Finds a node that satisfy the predicate and returns it
 */
public class Extensions {
    /**
     * The method finds the child node of root that satisfies the predicate
     *
     * @param root root node
     * @param predicate predicate to test against
     * @param getChildren
     * @param <T>
     * @return node that satisfies predicate
     */
    public static <T> T findWhere(final T root, final Predicate<T> predicate, final Function<T, Iterable<T>> getChildren) {
        // YOUR SOLUTION GOES HERE
        //Return the root if it that satisfy the predicate
        if (predicate.test(root)) {
            return root;
        }

        final Iterable<T> listOfChildren = getChildren.apply(root);

        //If root does not satisfy the predicate, check its children
        for (final T t : listOfChildren) {
            return findWhere(t, predicate, getChildren);
        }

        return null;
    }
}